Documentation/DesignGuidelines.md

```diff
@@ -6,28 +6,459 @@ Guidelines](http://blogs.msdn.com/b/rxteam/archive/2010/10/28/rx-design-guidelin
+there are few restrictions on what those values can mean or be used for, except
+that they must be of the same type.
+
+As an example, the value might represent an element from a collection, or
+a progress update about some long-running operation. The value of a `Next` event
+might even represent nothing at all—for example, it’s common to use a value type
+of `()` to indicate that something happened, without being more specific about
+what that something was.
+
+Most of the event stream [operators][] act upon `Next` events, as they represent the
+“meaningful data” of a signal or producer.
+
+#### Errors behave like exceptions and propagate immediately
+
+`Error` events indicate that something went wrong. Errors are fatal, and
+propagate as quickly as possible to the consumer for handling.
+
+Errors also behave like exceptions, in that they “skip” operators, terminating
+them along the way. In other words, most [operators][] immediately stop doing work
+when an error is received, and then propagate the error onward. This even
+applies to time-shifted operators, like [`delay`][delay]—which, despite its name, will
+forward any errors immediately.
+
+Consequently, errors should only be used to represent “abnormal” termination. If
+it is important to let operators (or consumers) finish their work, a `Next`
+event describing the result might be more appropriate.
+
+#### Completion indicates success
+
+An event stream sends `Completed` when the operation has completed successfully,
+or to indicate that the stream has terminated normally.
+
+Many operators manipulate the `Completed` event to shorten or extend the
+lifetime of an event stream.
+
+For example, [`take`][take] will complete after the specified number of values have
+been received, thereby terminating the stream early. On the other hand, most
+operators that accept multiple signals or producers will wait until _all_ of
+them have completed before forwarding a `Completed` event, since a successful
+outcome will usually depend on all the inputs.
+
+#### Interruption cancels outstanding work and usually propagates immediately
+
+An `Interrupted` event is sent when an event stream should cancel processing.
+Interruption is somewhere between [success](#completion-indicates-success)
+and [failure](#errors-behave-like-exceptions-and-propagate-immediately)—the
+operation was not successful, because it did not get to finish, but it didn’t
+necessarily “fail” either.
+
+Most [operators][] will propagate interruption immediately, but there are some
+exceptions. For example, the [flattening operators][flatten] will ignore
+`Interrupted` events that occur on the _inner_ producers, since the cancellation
+of an inner operation should not necessarily cancel the larger unit of work.
+
+RAC will automatically send an `Interrupted` event upon [disposal][Disposables], but it can
+also be sent manually if necessary. Additionally, [custom
```

Comment:
`Interrupted` is sort of the oddball that's easiest to forget. To help with that we should also encourage `switch`-ing over events (if not already) when implementing custom operators. That way the compiler can catch if you forget to forward a case which it'll miss when using the `observe(next: ..., error ...)` convenience method.